code/go/pkg/validator/validator_fields_test.go

```diff
@@ -0,0 +1,521 @@
+				},
+			},
+			expectedErrors: []string{
+				`field 0: Must not be present`,
+			},
+		},
+		{
+			title:           "runtime fields: invalid runtime value",
+			packageTemplate: "integration_v3_0",
+			fields: []map[string]any{
+				{
+					"name":    "runtime_field",
+					"type":    "boolean",
+					"runtime": 30,
+				},
+			},
+			expectedErrors: []string{
+				`field 0.runtime: Invalid type. Expected: string, given: integer`,
+			},
+		},
+		{
+			title:           "runtime fields: invalid runtime boolean",
+			packageTemplate: "integration_v3_0",
+			fields: []map[string]any{
+				{
+					"name":    "runtime_field",
+					"type":    "binary",
+					"runtime": true,
+				},
+			},
+			expectedErrors: []string{
+				`field 0: Must not be present`,
+			},
+		},
+		{
+
+			title:           "runtime fields: invalid runtime boolean",
+			packageTemplate: "integration_v3_0",
+			fields: []map[string]any{
+				{
+					"name":    "runtime_field",
+					"type":    "binary",
+					"runtime": "doc['message'].value().doSomething()",
+				},
+			},
+			expectedErrors: []string{
+				`field 0: Must not be present`,
+			},
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.title, func(t *testing.T) {
+			packagePath := createPackageWithFields(t, "testpackage", c.packageTemplate, c.fields)
+			err := ValidateFromPath(packagePath)
+			if len(c.expectedErrors) == 0 {
+				assert.NoError(t, err)
+				return
+			}
+			require.Error(t, err)
+
+			errs, ok := err.(specerrors.ValidationErrors)
+			require.True(t, ok)
+			assert.Len(t, errs, len(c.expectedErrors))
+
+			for _, foundError := range errs {
+				// Trim the part of the error that refers to the file where the error was found.
+				_, trimmedError, found := strings.Cut(foundError.Error(), ": ")
+				if assert.True(t, found) {
+					assert.Contains(t, c.expectedErrors, trimmedError)
+				}
+			}
+		})
+	}
+}
+
+func createPackageWithFields(t *testing.T, pkgName string, pkgTemplate string, fields any) string {
+	var encodedFields []byte
+	switch fields.(type) {
+	case nil:
+	case []byte:
+		encodedFields = fields.([]byte)
+	default:
+		var jsonFields bytes.Buffer
+		enc := yaml.NewEncoder(&jsonFields)
+		err := enc.Encode(fields)
+		require.NoError(t, err)
+		encodedFields = jsonFields.Bytes()
+	}
+
+	return createPackageWithRawFields(t, pkgName, pkgTemplate, encodedFields)
+}
+
+func createPackageWithRawFields(t *testing.T, pkgName string, pkgTemplate string, rawFields []byte) string {
+	basePackageDir := filepath.Join("testdata", "templates", pkgTemplate)
+	require.DirExists(t, basePackageDir, "checking if template package exists")
+
+	pkgDir := filepath.Join(t.TempDir(), pkgName)
+	err := cp.Copy(basePackageDir, pkgDir)
+	require.NoError(t, err, "copying template package")
+
+	if len(rawFields) > 0 {
+		fieldsPath := filepath.Join(pkgDir, "data_stream", "test", "fields", "fields.yml")
+		err := os.WriteFile(fieldsPath, rawFields, 0644)
+		require.NoError(t, err, "creating fields file")
+	}
```

Comment:
This looks really nice! 
Following this approach, it probably could also be added specific unit tests for some other packages tested in TestValidateFile from `validator_test.go`. For instance those test packages that are testing fields on `manifest.yml` (missing_version, bad_github_owner, bad_owner_type), data stream manifests, ingest pipelines...

Like having a base template package, and the test could add any file within the package with any content. Once that package is created call `ValidateFromPath()` as it is done here.

Do you think it would be possible to do that ? I don't mean to do it in this PR

Reply:
Agree, we could follow this approach in more cases. We can iterate later. Though many of these changes don't have so many cases and don't change so much as field, so maybe the investment doesn't have so many benefits.